server/src/test/java/org/opensearch/action/search/AbstractSearchAsyncActionTests.java

```diff
@@ -528,6 +550,544 @@ public void onFailure(Exception e) {
+                assertNotNull(context);
+                fetchPhaseEnd.incrementAndGet();
+            }
+
+            @Override
+            public void onExpandSearchPhaseStart(SearchPhaseContext context) {
+                assertNotNull(context);
+                expandPhaseStart.incrementAndGet();
+            }
+
+            @Override
+            public void onExpandSearchPhaseFailure(SearchPhaseContext context) {
+                assertNotNull(context);
+                expandPhaseFailure.incrementAndGet();
+            }
+
+            @Override
+            public void onExpandSearchPhaseEnd(SearchPhaseContext context, long tookTime) {
+                assertNotNull(context);
+                expandPhaseEnd.incrementAndGet();
+            }
+        };
+        final List<SearchRequestOperationsListener> requestOperationListeners = new ArrayList<>(Arrays.asList(testListener, testListener));
+
+        SearchQueryThenFetchAsyncAction action = createSearchQueryThenFetchAsyncAction();
+        action.setSearchListenerList(requestOperationListeners);
+
+        SearchDfsQueryThenFetchAsyncAction searchDfsQueryThenFetchAsyncAction = createSearchDfsQueryThenFetchAsyncAction();
+        searchDfsQueryThenFetchAsyncAction.setSearchListenerList(requestOperationListeners);
+
+        CanMatchPreFilterSearchPhase canMatchPreFilterSearchPhaseAction = createCanMatchPreFilterSearchPhase();
+        canMatchPreFilterSearchPhaseAction.setSearchListenerList(requestOperationListeners);
+
+        action.start();
+        assertEquals(2, queryPhaseStart.get());
+
+        FetchSearchPhase fetchPhase = createFetchSearchPhase();
+        ShardId shardId = new ShardId(randomAlphaOfLengthBetween(5, 10), randomAlphaOfLength(10), randomInt());
+        SearchShardIterator searchShardIterator = new SearchShardIterator(null, shardId, Collections.emptyList(), OriginalIndices.NONE);
+        searchShardIterator.resetAndSkip();
+        action.skipShard(searchShardIterator);
+        action.executeNextPhase(action, fetchPhase);
+
+        assertEquals(2, queryPhaseEnd.get());
+
+        searchDfsQueryThenFetchAsyncAction.start();
+        searchDfsQueryThenFetchAsyncAction.skipShard(searchShardIterator);
+        searchDfsQueryThenFetchAsyncAction.executeNextPhase(searchDfsQueryThenFetchAsyncAction, fetchPhase);
+
+        canMatchPreFilterSearchPhaseAction.start();
+
+        assertEquals(2, dfsPreQueryPhaseStart.get());
+        assertEquals(2, dfsPreQueryPhaseEnd.get());
+        assertEquals(2, canMatchPhaseStart.get());
+        assertEquals(4, fetchPhaseStart.get());
+
+        ExpandSearchPhase expandPhase = createExpandSearchPhase();
+        action.executeNextPhase(fetchPhase, expandPhase);
+        assertEquals(2, expandPhaseStart.get());
+
+        action.sendSearchResponse(InternalSearchResponse.empty(), null);
+    }
+
+    public void testMultiThreadCoordinateStats() throws InterruptedException {
```

Comment:
This Multi threading tests doesn't make sense when I think again.
You are creating your own - SearchRequestOperationsListener with just incrementAndGet which is known to be thread safe. And then verifying that. So we are not really testing coordinator stats.

So I am thinking we should add these multi threaded tests as part of SearchCoordinatorStatsTests. Call multiple such requests and trigger coordinator stats functions like onQueryPhaseEnd to see if it is storing took time and counter metric correctly.